pkg/ocmmanager/ocm_upgrade_config_manager.go

```diff
@@ -0,0 +1,439 @@
+	// Fetch the current UpgradeConfig instance, if it exists
+	upgradeConfigs := &upgradev1alpha1.UpgradeConfigList{}
+	err := c.List(context.TODO(), upgradeConfigs, &client.ListOptions{})
+	if err != nil {
+		return false, fmt.Errorf("unable to retrieve upgradeconfigs")
+	}
+
+	// If there are no policies available, or if there is no next upgrade scheduled, do not retain any UpgradeConfigs
+	if len(upgradePolicies.Items) == 0 || upgradePolicies.Items[0].NextRun == "" {
+		if len(upgradeConfigs.Items) > 0 {
+			for _, upgradeConfig := range upgradeConfigs.Items {
+				log.Info(fmt.Sprintf("Removing expired UpgradeConfig %s", upgradeConfig.Name))
+				err = c.Delete(context.TODO(), &upgradeConfig)
+				if err != nil {
+					return false, fmt.Errorf("unable to delete existing upgrade config: %v", err)
+				}
+			}
+		}
+		return true, nil
+	}
+
+	// Only handling singular upgrade policies for now - but log if that arises
+	if len(upgradePolicies.Items) > 1 {
+		log.Info("More than one Upgrade Policy received, only considering the first.")
+	}
+	upgradePolicy := upgradePolicies.Items[0]
+
+	// Set up an UpgradeConfig that reflects the policy
+	replacementUpgradeConfig := upgradev1alpha1.UpgradeConfig{}
+
+	// Check if we have an existing UpgradeConfig to compare against, for the refresh
+	originalUpgradeConfig := upgradev1alpha1.UpgradeConfig{}
+	if len(upgradeConfigs.Items) > 0 {
+		originalUpgradeConfig = upgradeConfigs.Items[0]
+		// If there was an existing UpgradeConfig, make a clone of its contents
+		originalUpgradeConfig.DeepCopyInto(&replacementUpgradeConfig)
+	} else {
+		// No existing UpgradeConfig exists, give the new one the default name/namespace
+		operatorNS, err := getOperatorNamespace()
+		if err != nil {
+			return false, fmt.Errorf("unable to determine running namespace, missing env OPERATOR_NAMESPACE")
+		}
+		replacementUpgradeConfig.Name = UPGRADECONFIG_CR_NAME
+		replacementUpgradeConfig.Namespace = operatorNS
+	}
+
+	// And build up the replacement UpgradeConfig based on the policy
+	// determine channel from channel group
+	upgradeChannel, err := inferUpgradeChannelFromChannelGroup(cluster.Version.ChannelGroup, upgradePolicy.Version)
+	if err != nil {
+		return false, fmt.Errorf("unable to determine channel from channel group '%v' and version '%v'", cluster.Version.ChannelGroup, upgradePolicy.Version)
+	}
+	replacementUpgradeConfig.Spec = upgradev1alpha1.UpgradeConfigSpec{
+		Desired: upgradev1alpha1.Update{
+			Version: upgradePolicy.Version,
+			Channel: *upgradeChannel,
+		},
+		UpgradeAt:            upgradePolicy.NextRun,
+		PDBForceDrainTimeout: int32(upgradePolicy.NodeDrainGracePeriod.Value),
+		Type:                 upgradev1alpha1.UpgradeType(upgradePolicy.UpgradeType),
+	}
+	replacementUpgradeConfig.Status = upgradev1alpha1.UpgradeConfigStatus{}
+
+	// is there a difference between the original and replacement?
+	changed := !reflect.DeepEqual(replacementUpgradeConfig.Spec, originalUpgradeConfig.Spec)
```

Comment:
There is a controllerruntime function that we could use?
https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

Reply:
I tried that, and it works, but it needs more time than I have available right now to properly work in a mocked test sense. It can be a future fix if anyone wants to.

Reply:
Added to the refactor jira